Fix 938 remove dangling exec containers

kodumbeats opened the pull request:
<!--
Thank you for sending the PR! We appreciate you spending the time to work on these changes.

Help us understand your motivation by explaining why you decided to make this change.

You can learn more about contributing to appwrite here: https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md

Happy contributing!

-->

## What does this PR do?

This PR adds a `delete` trigger to the functions worker to handle cleanup of dangling containers left from DELETE API endpoints. Currently, the containers remain unless they are bumped due to the 10 container cleanup.

## Test Plan

- Create function
- Deploy, activate, and execute a few tags
- Check for containers with `docker ps`
- Delete tag/function
- Check the container has been deleted

## Related PRs and Issues

https://github.com/appwrite/appwrite/issues/938

### Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?

Yes.

eldadfux commented:
Please sync 0.8/dev for tests to pass

eldadfux commented on app/controllers/api/functions.php:
```diff
@@ -417,6 +417,21 @@
             ->setParam('document', $function->getArrayCopy())
         ;
 
+        $tags = $projectDB->getCollection([
```
This is a loop we better manage inside the worker as it has the potential to contain unlimited amounts of tags. Use the deletes worker as a reference to view how we delete big lists of data using chunks rather than in one go.

kodumbeats commented:
What am I missing? Right when I get to the `delete` switch condition, I query for the function document, like every other trigger does, but this one throws an error: 
```
appwrite-worker-functions       | [notice] Starting work on (Job{v1-functions} | ID: 473e3626df9cf42f369c5a9cc5e0e6ed | FunctionsV1 | [{"functionId":"6075f50c86a9f","tagId":"6075f51769abf","trigger":"delete"}])
appwrite-worker-functions       | [PDO] MySQL connection restarted
appwrite-worker-functions       | [critical] (Job{v1-functions} | ID: 473e3626df9cf42f369c5a9cc5e0e6ed | FunctionsV1 | [{"functionId":"6075f50c86a9f","tagId":"6075f51769abf","trigger":"delete"}]) has failed PDOException: SQLSTATE[42S02]: Base table or view not found: 1146 Table 'appwrite.app_.database.documents' doesn't exist in /usr/src/code/src/Appwrite/Extend/PDOStatement.php:88
appwrite-worker-functions       | Stack trace:
appwrite-worker-functions       | #0 /usr/src/code/src/Appwrite/Extend/PDOStatement.php(88): PDOStatement->execute(NULL)
appwrite-worker-functions       | #1 /usr/src/code/src/Appwrite/Database/Adapter/MySQL.php(76): Appwrite\Extend\PDOStatement->execute()
appwrite-worker-functions       | #2 /usr/src/code/src/Appwrite/Database/Adapter/Redis.php(48): Appwrite\Database\Adapter\MySQL->getDocument('6075f50c86a9f')
appwrite-worker-functions       | #3 /usr/src/code/src/Appwrite/Database/Database.php(207): Appwrite\Database\Adapter\Redis->getDocument('6075f50c86a9f')
appwrite-worker-functions       | #4 /usr/src/code/app/workers/functions.php(273): Appwrite\Database\Database->getDocument('6075f50c86a9f')
appwrite-worker-functions       | #5 /usr/src/code/vendor/resque/php-resque/lib/Resque/Job.php(201): FunctionsV1->perform()
appwrite-worker-functions       | #6 /usr/src/code/vendor/resque/php-resque/lib/Resque/Worker.php(303): Resque_Job->perform()
appwrite-worker-functions       | #7 /usr/src/code/vendor/resque/php-resque/lib/Resque/Worker.php(242): Resque_Worker->perform(Object(Resque_Job))
appwrite-worker-functions       | #8 /usr/src/code/vendor/resque/php-resque/bin/resque(185): Resque_Worker->work(5, false)
appwrite-worker-functions       | #9 {main}
```

eldadfux commented:
> _What am I missing?_
> 
> I'm moving the database/event logic for removing functions into the functions worker because I need to query the function in chunks (to delete tags) before it can be removed from the database.
> 
> Right when I get to the `delete` switch condition, I query for the function document ([here](https://github.com/appwrite/appwrite/pull/1078/files#diff-16d93434c8aa21e9cfcdcffdf3228784e5d4801476c10df941dbdc2383a5b9d6R272-R274)), like every other trigger does, but this one throws an error:
> 
> ```
> appwrite-worker-functions       | [notice] Starting work on (Job{v1-functions} | ID: 473e3626df9cf42f369c5a9cc5e0e6ed | FunctionsV1 | [{"functionId":"6075f50c86a9f","tagId":"6075f51769abf","trigger":"delete"}])
> appwrite-worker-functions       | [PDO] MySQL connection restarted
> appwrite-worker-functions       | [critical] (Job{v1-functions} | ID: 473e3626df9cf42f369c5a9cc5e0e6ed | FunctionsV1 | [{"functionId":"6075f50c86a9f","tagId":"6075f51769abf","trigger":"delete"}]) has failed PDOException: SQLSTATE[42S02]: Base table or view not found: 1146 Table 'appwrite.app_.database.documents' doesn't exist in /usr/src/code/src/Appwrite/Extend/PDOStatement.php:88
> appwrite-worker-functions       | Stack trace:
> appwrite-worker-functions       | #0 /usr/src/code/src/Appwrite/Extend/PDOStatement.php(88): PDOStatement->execute(NULL)
> appwrite-worker-functions       | #1 /usr/src/code/src/Appwrite/Database/Adapter/MySQL.php(76): Appwrite\Extend\PDOStatement->execute()
> appwrite-worker-functions       | #2 /usr/src/code/src/Appwrite/Database/Adapter/Redis.php(48): Appwrite\Database\Adapter\MySQL->getDocument('6075f50c86a9f')
> appwrite-worker-functions       | #3 /usr/src/code/src/Appwrite/Database/Database.php(207): Appwrite\Database\Adapter\Redis->getDocument('6075f50c86a9f')
> appwrite-worker-functions       | #4 /usr/src/code/app/workers/functions.php(273): Appwrite\Database\Database->getDocument('6075f50c86a9f')
> appwrite-worker-functions       | #5 /usr/src/code/vendor/resque/php-resque/lib/Resque/Job.php(201): FunctionsV1->perform()
> appwrite-worker-functions       | #6 /usr/src/code/vendor/resque/php-resque/lib/Resque/Worker.php(303): Resque_Job->perform()
> appwrite-worker-functions       | #7 /usr/src/code/vendor/resque/php-resque/lib/Resque/Worker.php(242): Resque_Worker->perform(Object(Resque_Job))
> appwrite-worker-functions       | #8 /usr/src/code/vendor/resque/php-resque/bin/resque(185): Resque_Worker->work(5, false)
> appwrite-worker-functions       | #9 {main}
> ```

Our dependency injection doesn't work well for the workers, I'm pretty sure you just need to init the DB connection:
https://github.com/appwrite/appwrite/blob/078b5360d3a8ff1ec2cdfbd21189090324ab38cc/app/workers/deletes.php#L331-L358

kodumbeats commented:
@eldadfux Could you share your thoughts on these test failures?

To be able to batch delete tags, I moved the [database deletion step](https://github.com/appwrite/appwrite/pull/1078/files#diff-3c8abdcca7e143e051f729255409460b7fffbb41a3cf8dcc11c57651243662ecL411-L413) inside the functions worker, because I need the document to query tags in batches.  However, if you call `functions.get` immediately after `functions.delete` ([like we do in tests](https://github.com/appwrite/appwrite/blob/dev/tests/e2e/Services/Functions/FunctionsCustomServerTest.php#L427-L445)), you'll get a `200` response code, because the function is still in the database. 

To try to get tests to pass, I added a sleep in between the calls to give the functions worker time to process, but it doesn't feel good. Is there a more elegant solution to solve this problem? I'd rather not increase the sleep just to get tests to pass.
